### DISCUSSION TITLE
Installing the Teleport Slack plugin in Kubernetes

### DISCUSSION THREAD

**Discussion by webvictim** [gravitational/teleport#10974] [Show and tell]
## NOTE: New official Helm Chart

Since this post was written, Teleport has published an official Helm chart for the Slack plugin.

You can find this here: **https://github.com/gravitational/teleport-plugins/tree/master/charts/access/slack**

The legacy method will be preserved here for posterity.

### Legacy method

Here are some basic instructions on one way to install the Teleport Slack plugin in a Kubernetes cluster.

Caveats:
- This chart references a Docker image which is not built by Teleport CI and will not be updated on a regular basis, this is just a one-off solution.
  - Teleport is working on official versions of this chart and images to go along with it. There is no timeline for this currently.
  - You can build your own Docker image by following this discussion here: https://github.com/gravitational/teleport/discussions/8679
- The config file must be kept in the format which it's currently using (a quoted YAML scalar) otherwise the plugin will not work.
- This chart is not provided by Teleport itself, just an employee who wanted to experiment. Do not expect support from Teleport for the use of this chart; you are on your own.

Instructions:
1. Create the proper roles in your Teleport cluster by following the Slack docs on our docs page (https://goteleport.com/docs/enterprise/workflow/ssh-approval-slack/#create-user-and-role-for-access)

2. Generate an identity file (.pem file) on your Teleport auth server:
```bash
$ tctl auth sign --user=access-plugin —out=access-plugin.pem —ttl=8190h
```

(if Teleport is running inside Kubernetes, this may look more like this):
```bash
$ kubectl -n teleport exec deploy/teleport -t /bin/bash -c "tctl auth sign --user=access-plugin —out=access-plugin.pem —ttl=8190h"
$ kubectl -n teleport cp ./access-plugin.pem .
```

3. Download this zipped tarball file and extract it: [teleport-plugin-slack-9.1.0.zip](https://github.com/gravitational/teleport/files/8563738/teleport-plugin-slack-9.1.0.zip) (note that this is a `.tar.gz` file inside a `.zip` file due to Github's file upload restrictions, so you will need to extract both)

4. cd into the directory created after extraction
```bash
$ cd teleport-plugin-slack
```

5. Update your values.yaml file accordingly.
- Replace `teleport.example.com` with your Teleport hostname
- Replace the `xoxb...` token with your actual Slack token from the Slack website (see docs for configuration instructions)
- Replace `#general` with the channel name that you wish to send access requests to.
```yaml 
# Values you must set
proxyAddr: "teleport.example.com:443"

slack:
  token: "xoxb-96010..."
  # this should be provided as a scalar value due to TOML quoting constraints
  recipients: |
    ["#general"]
```

6. Execute kubectl command to create namespace
```bash
$ kubectl create ns teleport-plugin-slack
```

7. Install identity file (from the directory you saved the .pem file in earlier)
```bash
$ kubectl -n teleport-plugin-slack create secret generic teleport-plugin-slack-identity --from-file=../access-plugin.pem
```

8. Install the Helm chart
```
$ helm install -n teleport-plugin-slack teleport-plugin-slack . -f values.yaml
```

**Comment by sunghospark-calm** [comment]
Would love to get this implemented soon. For enterprise user these access request plugins are very essential part of the work flow and without this being implemented [the documentation](https://goteleport.com/docs/enterprise/workflow/ssh-approval-slack/?scope=enterprise) is basically not applicable for the users with k8s setup.

**Reply by sunghospark-calm** [reply]
I tried to follow the above steps. A quick feedback, please stay consistent with the file names.
The auth file was first mentioned as `auth.pem`, then it suddenly changes to `access-plugin.pem`
Then the secret the helm chart actually looks for is `teleport-slack-plugin-identity` instead of `teleport-plugin-slack-identity`

When everything was done, The container doesn't start up with following message:

```
ERRO   Terminating with fatal error... error:[
ERROR REPORT:
Original Error: *errors.errorString (9, 2): no value can start with l
Stack Trace:
	/go/src/github.com/gravitational/teleport-plugins/access/slack/config.go:60 main.LoadConfig
	/go/src/github.com/gravitational/teleport-plugins/access/slack/main.go:68 main.run
	/go/src/github.com/gravitational/teleport-plugins/access/slack/main.go:59 main.main
	/usr/local/go/src/runtime/proc.go:255 runtime.main
	/usr/local/go/src/runtime/asm_amd64.s:1581 runtime.goexit
User Message: (9, 2): no value can start with l] lib/bail.go:17
```

**Reply by webvictim** [reply]
This usually means that your TOML config isn't formatted correctly - it's quite finicky to represent TOML with YAML. It's most likely the `recipients` section. Can you share your values file?

**Comment by stevenGravy** [comment]
Helm charts for Access Plugins are now available at https://github.com/gravitational/teleport-plugins/tree/master/charts/access
